kinto/core/resource/__init__.py

```diff
@@ -460,17 +461,42 @@ def patch(self):
         existing = self._get_record_or_404(self.record_id)
         self._raise_412_if_modified(existing)
 
-        try:
-            # `data` attribute may not be present if only perms are patched.
-            changes = self.request.json.get('data', {})
-        except ValueError:
-            # If no `data` nor `permissions` is provided in patch, reject!
-            # XXX: This should happen in schema instead (c.f. ShareableViewSet)
-            error_details = {
-                'name': 'data',
-                'description': 'Provide at least one of data or permissions',
-            }
-            raise_invalid(self.request, **error_details)
+        content_type = str(self.request.headers.get('Content-Type'))
+        # patch is specified as a list of operations (RFC 6902)
+        if content_type == 'application/json-patch+json':
+            try:
+                ops = self.request.json
+                self.patch_changes = JsonPatch(ops, existing)
+
+                if '__permissions__' in existing:
+                    perms = existing['__permissions__']
+                    self.patch_changes.base_permissions.update(perms)
+
+                changes = self.patch_changes.apply_ops()
```

Comment:
This looks like you're applying the operations here, whereas we have a `apply_changes()` method.
I suggest that you keep the *flow*, here I would see something like:
```py
if content_type == 'application/json-patch+json':
    changes = self.request.json
else:
    try:
        changes = self.request.json.get('data', {})
        # ...
```